### DISCUSSION TITLE
SSH Connection Provider Package Setup

### DISCUSSION THREAD

**Discussion by bneal1987** [apache/airflow#19979] [Q&A]
We are having issues installing the SSH provider package. The 'pip install apache-airflow-providers-ssh==2.3.0' was successful, but we are unable to see the SSH connection in the admin > connections > add connection > conn type drop down menu.

Currently able to connect to airflow via the browser on local machine and through terminal server.

Can we provide any logs for further troubleshooting? If so, can you provide the path and steps for this as well?

Airflow is installed on Ubuntu 18.04 and currently does not have access to the internet so all changes are done locally, but on the same subnet as terminal server for connectivity. Able to RDP into the server as well to make local changes through a GUI interface.

sudo pip install --no-index --find-links=/usr/local/lib/python3.8/apache-airflow-providers-ssh-2.3.0 setup.py (this command did not throw any errors when running locally to try and complete the SSH provider package setup, still do not see it in the drop-down)

We are seeing the connections by using the following command: airflow connections list - This shows the SSH connection type being available, but unable to see it in the web interface when creating a new connection.

Thank you very much for your help!

**Comment by potiuk** [comment]
You likely did not restart your webserver (or your webserver does not have the package installed - depending on what your deployment is it might use different image/virtualenv). Re logs - you can find information about airflow logging architecture there  https://airflow.apache.org/docs/apache-airflow/stable/logging-monitoring/index.html?highlight=logging

**Reply by bneal1987** [reply]
To confirm, we restarted the webserver along with the actual server and getting the same results. Is there another way to restart the webserver? 

We installed airflow packages using pip/python. 

Any other suggestions?

**Reply by potiuk** [reply]
You can check by running `airflow providers hooks`

you should see ssh:

![image](https://user-images.githubusercontent.com/595491/144646936-a84f7c6b-0493-4951-b0a4-dac4d91ca1e8.png)

Then I got this (including SSH):

![Screenshot from 2021-12-03 18-33-14](https://user-images.githubusercontent.com/595491/144647118-31691b7c-d4a2-4682-96ac-2633bb12d2c6.png)

If you do not see it, then likely you installed the provider somewhere else - not in your airflow environment.

**Reply by bneal1987** [reply]
Wanted to provide some more notes

[follow up SSH error airflow 12072021.docx](https://github.com/apache/airflow/files/7670228/follow.up.SSH.error.airflow.12072021.docx)

Looking for any other suggestions.

**Reply by ghaynesPQS** [reply]
The results from the airflow providers hooks are below where you can see the SSH connection type.
![image](https://user-images.githubusercontent.com/97464384/148802481-b879fb3a-4cd9-401a-b013-3b1edcaa3c35.png)

However, when trying to add a connection through the GUI, the SSH does not appear.
![image](https://user-images.githubusercontent.com/97464384/148802726-4b86e9ad-69bd-409d-b3bd-55e1124bd4b7.png)

I also receive an error when trying to generate a URI to add the connection manually through CLI.
![image](https://user-images.githubusercontent.com/97464384/148802962-c5237770-0403-4e83-a48f-ada000756ca7.png)

The below image is a comparison of our install for the working sql_lite and the non working ssh.
![image](https://user-images.githubusercontent.com/97464384/148803564-27d782cf-1e8b-4a05-bcbd-30939291a313.png)

If anyone could offer guidance as to our next steps that would be much appreciated, Thanks!

**Reply by potiuk** [reply]
Can you please switch on the debug log for a while, restart airlfow webserver and see if there is an import error when SSH hook is imported? https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#logging-level

My best guess is that one ove the below imports generates an error

```
import os
import sys
import warnings
from base64 import decodebytes
from io import StringIO
from typing import Any, Dict, Optional, Sequence, Tuple, Type, Union

import paramiko
from paramiko.config import SSH_PORT
from sshtunnel import SSHTunnelForwarder
```

There is a potential problem we discovered (https://github.com/apache/airflow/issues/20709) that such missing imports might be currently hidden (turned into debug logs).

**Comment by bneal1987** [comment]
Thank you for the quick reply Jarek.

Just to make a note, this server does not have Internet access. Is that going to be an issue in what we are trying to achieve?

**Reply by potiuk** [reply]
Nope. Should not be.

**Comment by bneal1987** [comment]
Made the changes in settings.py in our airflow configuration directory and after reviewing the base log folder location and restarted the service, we are not seeing any of the logs.

There were no errors in the console output when we restarted the service.

Is there a specific location these logs are put for our review? Would like to review the information if we are getting errors are you suggested previously.

We can provide you some basic screenshots if need be.

Thanks,

Brandon Neal, CCPA
Infrastructure Engineer
PQS

**Comment by potiuk** [comment]
Should be just in webserver logs. and contain "Exception when importing" string (and you should see quite a number of DEBUG logs there) You can also upload the webserver log as a gist if you do not expect any sensitive information to be there.

**Reply by potiuk** [reply]
CAn you also check if the SSH is in the "Admin/Providers" tab in the UI?

**Comment by bneal1987** [comment]
Still not seeing SSH in the drop-down list for Connection type, local machine or accessing from a terminal server via web browser.

Will provide further information on logs after review.

**Comment by bneal1987** [comment]
We can close this conversation. 

Identified we had other issues related on why we were unable to get the SSH package installed correctly. Have re-built new server using Red Hat Enterprise Linux 8 with a new setup. 

Thanks,
